### PR TITLE
Configurable Warnings

### LINKED ISSUE #97: Duplicate #includes should not give a warning

### PR DESCRIPTION
**Pull request by spookydonut**
See included configREADME.md for usage

Closes #97

## Todo

- [x] make the toml file a cli switch
- [x] probably do is_disabled in some better fashion

### CODE REVIEW

**Comment by SpaceManiac** [comment]
I'm not in love with reshuffling Severity. What do you think about a design like:

```toml
[diagnostics]
should_call_parent = "warning"  # upgrades or downgrades severity
macro_redefined = "error"  # etc.
duplicate_includes = false  # disables entirely
```

**Comment by spookydonut** [comment]
> 
> 
> I'm not in love with reshuffling Severity. What do you think about a design like:
> 
> ```toml
> [diagnostics]
> should_call_parent = "warning"  # upgrades or downgrades severity
> macro_redefined = "error"  # etc.
> duplicate_includes = false  # disables entirely
> ```

That might work better yeah

File: src/dreammaker/config.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    undefine_undefined_macro: WarningLevel,
+
+    #[serde(default="Severity::default_all")]
+    error_level: Severity,
+}
+
+impl Warnings {
+    pub fn warning_level_for(&self, errortype: &'static str) -> WarningLevel {
+        match errortype {
+            "duplicate_include" => { return self.duplicate_includes },
```
Comment (SpaceManiac) [inline src/dreammaker/config.rs]: This whole match seems iffy, maybe we just want `warnings` to be a `HashMap<String, WarningLevel>` ?

File: src/dreammaker/config.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            _ => { return WarningLevel::Unset }
+        };
+    }
+
+    pub fn severe_enough(&self, severity: Severity) -> bool {
+        return severity <= self.error_level
+    }
+}
+
+pub fn read_config_toml(path: String) -> Config {
```
Comment (SpaceManiac) [inline src/dreammaker/config.rs]: Error handling is kind of a mess in this function, perhaps return a Result and let the caller deal with it

File: src/dreammaker/config.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?.to_lowercase();
+        let warninglevel = match s.as_str() {
+            "error" | "errors" => Ok(WarningLevel::Error),
+            "warning" | "warnings" => Ok(WarningLevel::Warning),
+            "info" | "infos" => Ok(WarningLevel::Info),
+            "hint" | "hints" => Ok(WarningLevel::Hint),
+            "false" | "off" => Ok(WarningLevel::Disabled),
+            _other => Ok(WarningLevel::Unset),
```
Comment (SpaceManiac) [inline src/dreammaker/config.rs]: Should probably return an `Err` so that a typo doesn't do nothing

File: src/dreammaker/config.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        };
+        warninglevel
+    }
+}
+
+impl From<WarningLevel> for Severity {
+    fn from(warninglevel: WarningLevel) -> Self {
```
Comment (SpaceManiac) [inline src/dreammaker/config.rs]: This `From` doesn't seem to be a complete conversion, perhaps the `match` where it's used should simply be made complete instead.

File: src/dreammaker/error.rs [hunk 5]
```diff
@@ -73,12 +76,14 @@ impl Context {
...
-                self.pretty_print_error(&mut stderr.lock(), &error)
-                    .expect("error writing to stderr");
-            }
+        if self.config.printable_error(&error) {
+            let stderr = termcolor::StandardStream::stderr(termcolor::ColorChoice::Auto);
+            self.pretty_print_error(&mut stderr.lock(), &error)
+                .expect("error writing to stderr");
+        }
+        // ignore errors with severity above configured level
+        if !self.config.registerable_error(&error) {
```
Comment (SpaceManiac) [inline src/dreammaker/error.rs]: I don't think I'm actually seeing the ability to change the severity of a diagnostic

**Pull request merged**